# Associate Rule Learning - Apriori Model (Grocery_Store Dataset)
# Strategy : To Create Set of Rules by finding out group of Customers who bought 'A' also bought 'B'. (Recommendation System)

## Data Preprocessing 

In [1]:
!pip install apyori

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
#header=none since we dont have the name of columns we will specify it 
dataset=pd.read_csv(r'E:\Project\Associate Rule Learning\Apriori and Eclat\Dataset\Market_Basket_Optimisation.csv', header=None)
#Instead of having the format in pandas dataframe we gonna have the data set in list of transactions
transactions= []
#For loop to populate this list ; 
#use append function to build the list
for i in range(0,7501): # :-Cloumn
    transactions.append([str(dataset.values[i,j])for j in range(0,20)]) #Single row for loop; #str - to make all string values for apyori model to learn

## Building and Training Apriori Model on Dataset 

In [4]:
from apyori import apriori
rules = apriori(transactions=transactions, min_support = 0.003, min_confidence = 0.2, min_lift = 3, min_length = 2, max_length = 2)
#Support = products that are repeated 3 times in a day ; 3x7=21 ; 21/7501 =0.00279
#Confidence = 0.8 - Few Rules /2 , 0.4 - Again Few Rules /2, 0.2 - good Amt. of rules
#Lift = 3(4,5,6,7,8,9) (Value = 3 takeb based on experimentation )
#min_length,max_lenth =2 (Since we want to find relation between a and b i.e buy this and get other for free)

## Visualiazing Results 

### Displaying Results Coming Directly from  Apriori Model

In [5]:
results=list(rules) #To get list of rules 
results

[RelationRecord(items=frozenset({'chicken', 'light cream'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]),
 RelationRecord(items=frozenset({'escalope', 'mushroom cream sauce'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)]),
 RelationRecord(items=frozenset({'escalope', 'pasta'}), support=0.005865884548726837, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)]),
 RelationRecord(items=frozenset({'honey', 'fromage blanc'}), support=0.003332888948140248, ordered_statistics=[OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confidence=0

#### item - one is people buy "light cream" are more likely to buy "Chicken" ; #Lift =4.84 (For Statement 1)

### Putting results in a well organised Pandas Dataframe (Unsorted) :

In [6]:
def inspect(results):
    lhs         = [tuple(result[2][0][0])[0] for result in results]
    rhs         = [tuple(result[2][0][1])[0] for result in results]
    supports    = [result[1] for result in results]
    confidences = [result[2][0][2] for result in results]
    lifts       = [result[2][0][3] for result in results]
    return list(zip(lhs, rhs, supports, confidences, lifts))
resultsinDataFrame = pd.DataFrame(inspect(results), columns = ['Left Hand Side', 'Right Hand Side', 'Support', 'Confidence', 'Lift'])

In [7]:
resultsinDataFrame

Left Hand Side Right Hand Side   Support  Confidence      Lift
0           light cream         chicken  0.004533    0.290598  4.843951
1  mushroom cream sauce        escalope  0.005733    0.300699  3.790833
2                 pasta        escalope  0.005866    0.372881  4.700812
3         fromage blanc           honey  0.003333    0.245098  5.164271
4         herb & pepper     ground beef  0.015998    0.323450  3.291994
5          tomato sauce     ground beef  0.005333    0.377358  3.840659
6           light cream       olive oil  0.003200    0.205128  3.114710
7     whole wheat pasta       olive oil  0.007999    0.271493  4.122410
8                 pasta          shrimp  0.005066    0.322034  4.506672

### Displaying Sorted Results (Descending Order on basis of there Lift): 

In [8]:
resultsinDataFrame.nlargest(n=10, columns='Lift')

Left Hand Side Right Hand Side   Support  Confidence      Lift
3         fromage blanc           honey  0.003333    0.245098  5.164271
0           light cream         chicken  0.004533    0.290598  4.843951
2                 pasta        escalope  0.005866    0.372881  4.700812
8                 pasta          shrimp  0.005066    0.322034  4.506672
7     whole wheat pasta       olive oil  0.007999    0.271493  4.122410
5          tomato sauce     ground beef  0.005333    0.377358  3.840659
1  mushroom cream sauce        escalope  0.005733    0.300699  3.790833
4         herb & pepper     ground beef  0.015998    0.323450  3.291994
6           light cream       olive oil  0.003200    0.205128  3.114710